Classes/Private/Manager/Video/OCTVideoEngine.m

```diff
@@ -0,0 +1,246 @@
+//
+//  OCTVideoEngine.m
+//  objcTox
+//
+//  Created by Chuong Vu on 6/19/15.
+//  Copyright (c) 2015 dvor. All rights reserved.
+//
+
+#import "OCTVideoEngine.h"
+#import "OCTVideoView.h"
+#import "DDLog.h"
+
+#undef LOG_LEVEL_DEF
+#define LOG_LEVEL_DEF LOG_LEVEL_VERBOSE
+
+@import AVFoundation;
+
+static uint8_t *reusableUChromaPlane;
+static uint8_t *reusableVChromaPlane;
+
+static const OSType kPixelFormat = kCVPixelFormatType_420YpCbCr8BiPlanarVideoRange;
+
+@interface OCTVideoEngine () <AVCaptureVideoDataOutputSampleBufferDelegate>
+
+@property (nonatomic, strong) AVCaptureSession *captureSession;
+@property (nonatomic, strong) AVCaptureVideoDataOutput *dataOutput;
+@property (nonatomic, strong) dispatch_queue_t processingQueue;
+@property (nonatomic, strong) OCTVideoView *videoView;
+
+@property (nonatomic, assign) NSUInteger sizeOfChromaPlanes;
+
+@end
+
+@implementation OCTVideoEngine
+
+- (instancetype)init
+{
+    DDLogVerbose(@"%@: init", self);
+    self = [super init];
+    if (! self) {
+        return nil;
+    }
+
+    _captureSession = [AVCaptureSession new];
+    _captureSession.sessionPreset = AVCaptureSessionPreset1280x720;
+    _dataOutput = [AVCaptureVideoDataOutput new];
+    _processingQueue = dispatch_queue_create("me.dvor.objcTox.OCTVideoEngineQueue", NULL);
+
+    return self;
+}
+
+#pragma mark - Public
+
+- (BOOL)setupWithError:(NSError **)error
+{
+    DDLogVerbose(@"%@: setupWithError", self);
+    AVCaptureDevice *videoCaptureDevice = [self frontCamera];
+    AVCaptureDeviceInput *videoInput = [AVCaptureDeviceInput deviceInputWithDevice:videoCaptureDevice error:error];
+
+    if (! videoInput) {
+        return NO;
+    }
+
+    [self.captureSession addInput:videoInput];
+
+    [self.dataOutput setAlwaysDiscardsLateVideoFrames:YES];
+    NSNumber *value  = [NSNumber numberWithUnsignedInt:kCVPixelFormatType_420YpCbCr8BiPlanarVideoRange];
+    [self.dataOutput setVideoSettings:[NSDictionary dictionaryWithObject:value
+                                                                  forKey:(id)kCVPixelBufferPixelFormatTypeKey]];
+    [self.dataOutput setSampleBufferDelegate:self queue:self.processingQueue];
+
+    [self.captureSession addOutput:self.dataOutput];
+
+    return YES;
+}
+
+- (void)startSendingVideo
+{
+    DDLogVerbose(@"%@: startSendingVideo", self);
+    self.processIncomingVideo = YES;
+    if ([self isVideoSessionRunning]) {
+        return;
+    }
```

Comment:
Yea there is no mention about it being a blocking call so I think we are safe there.